### ISSUE TITLE
False positive box_default inside macro

### ISSUE THREAD

**Issue by synek317** [rust-lang/rust-clippy#11868]
### Summary

If `Box::new` is used in a macro and then this macro is called with a default value at least once, then the lint is triggered even if in other macro invocations, `default` could not be used - see code below.

Link to the playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9e737f21f166c1cb8ad0cce037fb5649

### Lint Name

box_default

### Reproducer

I tried this code:

```rust
fn foo(_: &mut Vec<usize>) {}

macro_rules! bar {
    ($baz:expr) => {
        Box::leak(Box::new($baz))
    };
}

fn main() {
    foo(bar!(vec![]));
    foo(bar!(vec![1]));
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: `Box::new(_)` of default value
  --> src/main.rs:5:19
   |
5  |         Box::leak(Box::new($baz))
   |                   ^^^^^^^^^^^^^^ help: try: `Box::default()`
...
10 |     foo(bar!(vec![]));
   |         ------------ in this macro invocation
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#box_default
   = note: `#[warn(clippy::box_default)]` on by default
   = note: this warning originates in the macro `bar` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `playground` (bin "playground") generated 1 warning (run `cargo clippy --fix --bin "playground"` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.64s
```

I expected to see this happen:

```
    Checking playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.35s
```

### Version

```text
1.76.0-nightly (2023-11-24 37b2813a7be580ef5904)
```
